libbpf-tools/biolatency.c

```diff
@@ -0,0 +1,351 @@
+	int times;
+	bool timestamp;
+	bool queued;
+	bool per_disk;
+	bool per_flag;
+	bool milliseconds;
+	bool verbose;
+} env = {
+	.interval = 99999999,
+	.times = 99999999,
+};
+
+static volatile bool exiting;
+
+const char *argp_program_version = "biolatency 0.1";
+const char *argp_program_bug_address = "<ethercflow@gmail.com>";
+const char argp_program_doc[] =
+"Summarize block device I/O latency as a histogram.\n"
+"\n"
+"USAGE: biolatency [-h] [-T] [-m] [-Q] [-D] [-F] [interval] [count]\n"
+"\n"
+"EXAMPLES:\n"
+"    biolatency              # summarize block I/O latency as a histogram\n"
+"    biolatency 1 10         # print 1 second summaries, 10 times\n"
+"    biolatency -mT 1        # 1s summaries, milliseconds, and timestamps\n"
+"    biolatency -Q           # include OS queued time in I/O time\n"
+"    biolatency -D           # show each disk device separately\n"
+"    biolatency -F           # show I/O flags separately\n"
+"    biolatency -d sdc       # Trace sdc only\n";
+
+static const struct argp_option opts[] = {
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{ "timestamp", 'T', NULL, 0, "Include timestamp on output" },
+	{ "milliseconds", 'm', NULL, 0, "Millisecond histogram" },
+	{ "queued", 'Q', NULL, 0, "Include OS queued time in I/O time" },
+	{ "disk", 'D', NULL, 0, "Print a histogram per disk device" },
+	{ "flag", 'F', NULL, 0, "Print a histogram per set of I/O flags" },
+	{ "disk",  'd', "DISK",  0, "Trace this disk only" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'h':
+		argp_usage(state);
+		break;
+	case 'm':
+		env.milliseconds = true;
+		break;
+	case 'Q':
+		env.queued = true;
+		break;
+	case 'D':
+		env.per_disk = true;
+		break;
+	case 'F':
+		env.per_flag = true;
+		break;
+	case 'd':
+		env.disk = arg;
+		env.disk_len = strlen(arg) + 1;
+		if (env.disk_len > DISK_NAME_LEN) {
+			fprintf(stderr, "invaild disk name: too long\n");
+			argp_usage(state);
+		}
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static void print_cmd_flags(int cmd_flags)
```

Comment:
have you consider doing something like this instead:
```
static struct { int bit; const char * str; } flags[] = {
  { __REQ_NOWAIT, "NoWait-" },
  { __REQ_BACKGROUND, "Background-"},
  ...
};
for (i = 0; i < ARRAY_SIZE(flags); i++) {
    if (cmd_flags & flags[i].bit)
        printf(flags[i].str);
}
```

For REQ_OP_MASK similar approach:
```
static const char *op[] = {
  [REQ_OP_READ] = "Read",
  [REQ_OP_WRITE] = "Write",
  ...
};

if ((cmd_flags & REQ_OP_MASK) < ARRAY_SIZE(op)) {
    printf("Unknown");
else
    printf(op[cmd_flags & REQ_OP_MASK]);
}
```

It will be a bit nicer, IMO.

Reply:
Thanks, it looks much nicer.

Reply:
Fixed